### ISSUE TITLE
Extras using the same namespace/extras core folder

### ISSUE THREAD

**Issue by Jako** [modxcms/revolution#12879]
### Summary
What happens if two extras use the same namespace and/or the same extras core/assets folder.

### Step to reproduce/Observed behavior
I have not tried that, but there are signs, that both extras are installed without any message and the last installed extra will win.

### Expected behavior
There should be at least some check and a warning from the installer, before this issue could occur.

### Environment
MODX 2.4.2

**Comment by goldsky** [comment]
Logically, it'll be counted as an updated.
When extra developer submits the extra to MODX repo, extra will check the duplication.
So having the same exact namespace seems avoidable.

**Comment by Jako** [comment]
There is not only one repository available.

**Comment by inreti** [comment]
Additionally, if you have a non-extra component (e.g. something you created for your webapp) placed in [assets|core]/components/, you might have placed it into a folder whichs name is also a namespace of an extra. If you then install that extra, your custom component will be overwriten/broken, maybe without any visible warning for you. The installer should _not_ install the extra in that case! Instead, you should be forced to fix the issue in order to be able to install anything.

**Comment by OptimusCrime** [comment]
Is this a real problem? Does this need to be solved in a way? I can see the theoretical problems with same namespaces, but I fail to see the importance of solving this. Extras should have a unique namespace indicator.

**Comment by Alroniks** [comment]
And it also may be a good practice to check it via validators form developer of an extra side instead of locking it from MODX. Also, one case that now used in payment systems components for minishop, when extra adds a special class to minishop2 component folder in core/components. In this case, strict mode from MODX will break such ability.

**Comment by OptimusCrime** [comment]
So this would actually break functionality that is in use today?

**Comment by Alroniks** [comment]
Yes, it may. A lot of solutions for minishop2 are injecting own code into this folder https://github.com/bezumkin/miniShop2/tree/master/core/components/minishop2/custom. Of course, these components have the own namespaces for system settings and so on but depends on implementation it may break current state.

**Comment by OptimusCrime** [comment]
Thoughts, @Jako ?

**Comment by Jako** [comment]
Each provider should save and send an unique id for the extra. That would be saved on client side during the installation. That way those two extras with the same namespace could be identified and the second one could not be installed.

**Comment by OptimusCrime** [comment]
I don't follow. What is the unique id? Is that the namespace? Read @Alroniks's use case above. It would deny those types of extras.

**Comment by Omeryl** [comment]
@Jako And what if you switch providers? I don't think that solves the problem here.

**Comment by Jako** [comment]
@OptimusCrime @Omeryl The initial problem is the following: at the moment you could install packages with the same name namespace (= same folder in assets/components/ or core/components). This could happen with an own built package and a provider one or with two packages of of different providers.

If each provider (or package builder) would generate an unique id for an extra (that does not change with the extra versions) and that would be sent together with the package download, the MODX installation could compare this unique id with the id of the installed extra and stop the installation (or make it possible after a warning). That way we don't need a global namespace register.

**Comment by Mark-H** [comment]
If you're trusting different package providers to create globally unique IDs, why not trust that same package provider to have a unique package name? 

I think this is mostly a theoretical problem that is easily solved, if/when it happens, by discussing it among developers that have extras with the same name(space). A few years ago there were 2 extras that both used "analytics", that was also resolved quite quickly.

**Comment by Jako** [comment]
But what about a local generated/installed (maybe a patched) package that is overwritten by a package provider one? I just want to higher the level that someone can't do anything wrong.

**Comment by Mark-H** [comment]
How do you know if it is or isn't the goal of the person to install a newer version from the package provider? 

I can see theoretical scenarios where people might accidentally overwrite files, but all of them rely on either two extras with the same namespace (which can be avoided outside of the code) or the admin user taking complicated steps (like building a patched package) to cause the problem in the first place. Doesn't really seem worth spending days laboring over to get it working right for me.
